### PR TITLE
Respond to keepalive messages within a TCP socket

### PR DESCRIPTION
**Pull request by clemahieu**

### CODE REVIEW

File: nano/node/bootstrap.cpp [hunk 1]
```diff
@@ -2178,6 +2186,28 @@ void nano::bootstrap_server::receive_bulk_pull_account_action (boost::system::er
...
+void nano::bootstrap_server::receive_keepalive_action (boost::system::error_code const & ec, size_t size_a, nano::message_header const & header_a)
+{
+	if (!ec)
+	{
+		auto error (false);
+		nano::bufferstream stream (receive_buffer->data (), header_a.payload_length_bytes ());
+		std::unique_ptr<nano::keepalive> request (new nano::keepalive (error, stream, header_a));
+		if (!error)
+		{
+			add_request (std::unique_ptr<nano::message> (request.release ()));
```
Comment (wezrule) [inline nano/node/bootstrap.cpp]: can't this just be: `add_request (std::move (request));` ?

File: nano/core_test/network.cpp [hunk 2]
```diff
@@ -1392,3 +1392,31 @@ TEST (bulk_pull_account, basics)
...
+	socket->async_connect (system.nodes[0]->bootstrap.endpoint (), [&input, socket](boost::system::error_code const & ec) {
+		ASSERT_FALSE (ec);
+		socket->async_write (input, [&input](boost::system::error_code const & ec, size_t size_a) {
+			ASSERT_FALSE (ec);
+			ASSERT_EQ (input->size (), size_a);
+		});
+	});
+
+	auto output (keepalive.to_bytes ());
+	bool done (false);
```
Comment (SergiySW) [inline nano/core_test/network.cpp]: Probably should be atomic ?

File: nano/node/bootstrap.cpp (hunk 1 above)
Comment (clemahieu) [reply nano/node/bootstrap.cpp]: Yes, a lot of these need to be fixed to use c++13 semantics.

File: nano/node/bootstrap.cpp [hunk 3]
```diff
@@ -2235,9 +2265,35 @@ class request_response_visitor : public nano::message_visitor
...
+		connection->node->stats.inc (nano::stat::type::message, nano::stat::detail::keepalive, nano::stat::dir::in);
+		connection->node->network.merge_peers (message_a.peers);
+		nano::keepalive message;
+		connection->node->peers.random_fill (message.peers);
+		auto bytes = message.to_bytes ();
+		if (connection->node->config.logging.network_keepalive_logging ())
+		{
+			BOOST_LOG (connection->node->log) << boost::str (boost::format ("Keepalive req sent to %1%") % connection->socket->remote_endpoint ());
+		}
+		connection->socket->async_write (bytes, [connection = connection](boost::system::error_code const & ec, size_t size_a) {
```
Comment (SergiySW) [inline nano/node/bootstrap.cpp]: connection = connection (?)

Comment (wezrule) [reply nano/node/bootstrap.cpp]: connection is a member variable, so if you want to pass it in without passing `this` this is the way to do it (and [&var = var] for reference)

**Pull request merged**